_posts/2023-01-26-how-kani-helped-find-bugs-in-hifitime.md

```diff
@@ -0,0 +1,54 @@
+
+<img src="{{site.baseurl | prepend: site.url}}/assets/images/tai-vs-et-vs-tdb.png" alt="TT is fixed to TAI but TDB and ET are periodic" />
+
+
+## Avoiding loss of precision
+
+The precision of floating point values is constrained by its representation in the processor. Moreover, some microcontrollers do not have a floating point unit (FPU), meaning that any floating point operation is emulated through software, adding a considerable number of CPU cycles for any computation.
+
+Dates, or "epochs," are _simply_ the duration past a certain reference epoch. Operations on these epochs, like adding 26.7 days, will cause rounding errors (even sometimes in a single operations like on this [`f32`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b760579f103b7192c20413ebbe167b90)).
+
+Time scales have different reference epochs, and to correctly convert between different time scales, we must ensure that we don't lose any precision in the computation of these offsets. That's especially important given that one of the most common scales in astronomy is the Julian Date, whose reference epoch is 4713 BC (or -4712). Storing such a number with an `f64` would lead to very large rounding errors.
+
+Hence, to avoid loss of precision, hifitime stores all durations as a tuple of the number of centuries and the number of nanoseconds into this century. For example `(0, 1)` is a duration of one nanosecond. Then, the Epoch is a structure containing the TAI duration since the TAI reference epoch of 01 January 1900 at noon and the time scale in which the Epoch was initialized. ([_Maybe_ hifitime should be even more precise?](https://github.com/nyx-space/hifitime/issues/186))
+
+## Importance of Kani in hifitime
+
+The purpose of hifitime is to convert between time scales, and these drift apart from each other. Time scales are only used for scientific computations, so it's important to correctly compute the conversion between these time scales. As previously discussed, for ensure exactly one nanosecond precision, hifitime stores durations as a tuple of integers, as these aren't affected by successive rounding errors. The drawback of this approach is that we now have to rewrite the arithmetics of durations and get it right!
```

Comment:
We have to rewrite the arithmetics of durations on this (centuries, nanosecond) encoding and guarantee it satisfies expected properties.